stackrox/stackrox
# ROX-11101: Remove deprecated resources

## Description

The set resources used for access control is being simplified. A number of resources has been deprecated.

The goal here is to get rid of the deprecated resources, both in the database and in the codebase (UI, QA, central, DB). This PR focuses on the database content cleanup. Full PR chain:
https://github.com/stackrox/stackrox/pull/3036
https://github.com/stackrox/stackrox/pull/3115
https://github.com/stackrox/stackrox/pull/3116
https://github.com/stackrox/stackrox/pull/3428

## Checklist
~~- [ ] Determined and documented upgrade steps~~
~~- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~~

If any of these don't apply, please comment below.

## Testing Performed

Added automated tests for the migration.

Manual testing
- Deploy locally a 3.72.0 `central` in RocksDB mode.
- Manually create permission sets using the UI.
- Create permission sets using the API.
example:
```
curl --insecure -XPOST -u admin:$admin_password https://localhost:8000/v1/permissionsets \
--data '{"id":"","name":"OneAccessSubAPIPermissionSet", "description":"", "resourceToAccess":{"AuthProvider": "READ_ACCESS"}}'
```
- Check in the UI all created permission sets are present and contain the expected resource access.
- Check using the API that the created permission sets are present and contain the expected resource access.
example:
```
curl --insecure -XGET -u admin:$admin_password https://localhost:8000/v1/permissionsets
```
- Update the image tag in the deployment to `3.72.x-533-g1614bca3ad`.
- Check the migrator run in the central deployment logs, make sure the migration from sequence number `110` to `111` happened, and that the database sequence number at the end is at least `111`.
```
% kubectl logs -n stackrox deploy/central
No certificates found in /usr/local/share/ca-certificates
No certificates found in /etc/pki/injected-ca-trust
Migrator: 2022/11/07 12:44:40.057224 log.go:13: Info: In migrator.run()
pkg/migrations: 2022/11/07 12:44:40.077194 migration_version.go:58: Info: Migration version of database at /var/lib/stackrox/current: &{/var/lib/stackrox/current 3.72.1 108 0001-01-01 00:00:00 +0000 UTC}
clone/rocksdb: 2022/11/07 12:44:40.077311 db_clone_manager_impl.go:66: Info: Found clone current -> .db-init
clone/rocksdb: 2022/11/07 12:44:40.077364 db_clone_manager_impl.go:124: Info: Database clones:
clone/rocksdb: 2022/11/07 12:44:40.077387 db_clone_manager_impl.go:126: Info: current -> &{/var/lib/stackrox/current 3.72.1 108 0001-01-01 00:00:00 +0000 UTC}
clone/rocksdb: 2022/11/07 12:44:40.077523 db_clone_manager_impl.go:302: Info: Central has space to create backup for rollback: true, required: 39846073, available: 105016770560 with 0.200000 margin
clone/rocksdb: 2022/11/07 12:44:40.077560 db_clone_manager_impl.go:179: Info: Database rollback enabled. Copying database files and migrate it to current version.
pkg/migrations: 2022/11/07 12:44:40.530951 migration_version.go:58: Info: Migration version of database at /var/lib/stackrox/.db-411092ad-89a8-401a-837e-a55584fbcf75: &{/var/lib/stackrox/.db-411092ad-89a8-401a-837e-a55584fbcf75 3.72.1 108 0001-01-01 00:00:00 +0000 UTC}
Migrator: 2022/11/07 12:44:40.531016 log.go:18: Info: Clone to Migrate "temp", ""
Migrator: 2022/11/07 12:44:40.531033 log.go:13: Info: starting DB compaction
Migrator: 2022/11/07 12:44:40.531106 log.go:18: Info: Free fraction of 0.0938 (24576/262144) is < 0.7500. Will not compact
Migrator: 2022/11/07 12:44:41.056613 log.go:18: Info: In runner.Run
Migrator: 2022/11/07 12:44:41.057388 log.go:18: Info: Found DB at version 108, which is less than what we expect (111). Running migrations...
Migrator: 2022/11/07 12:44:41.067179 log.go:18: Info: Successfully updated DB from version 108 to 109
Migrator: 2022/11/07 12:44:41.073678 log.go:18: Info: Successfully updated DB from version 109 to 110
Migrator: 2022/11/07 12:44:41.075925 log.go:18: Info: Successfully updated DB from version 110 to 111
clone/rocksdb: 2022/11/07 12:44:41.076798 db_clone_manager_impl.go:212: Info: Persisting upgraded clone: temp
pkg/mtls/certwatch: 2022/11/07 12:44:41.328036 certwatch.go:61: Info: TLS certificate loaded, using the following cert for HTTPS: (SerialNumber: 424012542049970283456843807461678417531196, Subject: CN=*.demo.stackrox.com, DNSNames, [*.demo.stackrox.com]), watch dir: "/run/secrets/stackrox.io/default-tls-cert"
main: 2022/11/07 12:44:41.391397 main.go:247: Info: Running StackRox Version: 3.72.x-533-g1614bca3ad
version: 2022/11/07 12:44:41.413072 ensure.go:55: Info: Version found in the DB was current. We're good to go!
...

% kubectl exec -n stackrox deploy/central -- cat /var/lib/stackrox/current/migration_version.yaml
image: 3.72.x-533-g1614bca3ad
database: 111
lastpersisted: 0001-01-01T00:00:00Z
```
- After the migration, check that all the default and created permission sets still exist.
- After the migration, check the content of permission sets. The replaced permissions should not be there anymore, the replacing permissions should have the lowest access level across the replaced permissions.
- After the migration, check that roles can reference the permission sets in the UI.
- After the migration, check that roles can reference the permission sets using the API.
